samples/attested_tls/server/enc/openssl/server.cpp

```diff
@@ -0,0 +1,377 @@
+exit:
+    return ret;
+}
+
+void init_openssl_library()
+{
+    OpenSSL_add_ssl_algorithms();
+    SSL_load_error_strings();
+    ERR_load_crypto_strings();
+    SSL_load_error_strings();
+}
+
+int initalize_ssl_context(SSL_CTX*& ctx)
+{
+    int ret = -1;
+    const SSL_METHOD* method;
+    if ((ctx = SSL_CTX_new(SSLv23_server_method())) == NULL)
+    {
+        printf(TLS_SERVER " unable to create a new SSL context\n");
+        goto exit;
+    }
+    // choose TLSv1.2 by excluding SSLv2, SSLv3 ,TLS 1.0 and TLS 1.1
+    SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv2);
+    SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv3);
+    SSL_CTX_set_options(ctx, SSL_OP_NO_TLSv1);
+    SSL_CTX_set_options(ctx, SSL_OP_NO_TLSv1_1);
+    ret = 0;
+exit:
+    return ret;
+}
+
+int load_ssl_certificates_and_keys(SSL_CTX* ctx, X509*& cert, EVP_PKEY*& pkey)
+{
+    int ret = -1;
+    SSL_CTX_set_ecdh_auto(ctx, 1);
+    uint8_t* output_cert = NULL;
+    size_t output_cert_size = 0;
+    uint8_t* private_key_buf = NULL;
+    size_t private_key_buf_size = 0;
+    const unsigned char* cert_buf_ptr = NULL;
+
+    if (generate_certificate_and_pkey_buffers(
+            output_cert,
+            output_cert_size,
+            private_key_buf,
+            private_key_buf_size) == OE_FAILURE)
+    {
+        printf(" failed generating certificate and private key buffers \n");
+        goto exit;
+    }
+
+    // temporary buffer required as if d2i_x509 call is successful cert_buf_ptr
+    // is incremented to the byte following the parsed data. sending
+    // cert_buf_ptr as argument will keep output_cert pointer undisturbed.
+    cert_buf_ptr = output_cert;
+
+    if ((cert = d2i_X509(NULL, &cert_buf_ptr, (long)output_cert_size)) == NULL)
+    {
+        printf(TLS_SERVER
+               "Failed to convert DER fromat certificate to X509 structure\n");
+        goto exit;
+    }
+
+    if (!SSL_CTX_use_certificate(ctx, cert))
+    {
+        printf(TLS_SERVER "Cannot load certificate on the server\n");
+        goto exit;
+    }
+
+    if ((pkey = PEM_read_bio_PrivateKey(
+             BIO_new_mem_buf((void*)private_key_buf, -1), NULL, 0, NULL)) ==
+        NULL)
+    {
+        printf(TLS_SERVER
+               "Failed to convert private key buffer into EVP_KEY format\n");
+        goto exit;
+    }
+
+    if (!SSL_CTX_use_PrivateKey(ctx, pkey))
+    {
+        printf(TLS_SERVER "Cannot load private key on the server\n");
+        goto exit;
+    }
+
+    /* verify private key */
+    if (!SSL_CTX_check_private_key(ctx))
+    {
+        printf(TLS_SERVER
+               "Private key does not match the public certificate\n");
+        goto exit;
+    }
+
+    ret = 0;
+exit:
+    cert_buf_ptr = NULL;
+    oe_free_key(private_key_buf, private_key_buf_size, NULL, 0);
+    oe_free_attestation_certificate(output_cert);
+    return ret;
+}
+
+int create_listner_socket(int port, int& server_socket)
+{
+    int ret = -1;
+    struct sockaddr_in addr;
+    addr.sin_family = AF_INET;
```

Comment:
I'm sad to see IPv4-only samples.  
This is not a blocker, but I'd really like all samples to be IP-version agnostic.

Reply:
I could not understand complete implementation details of `create_socket `in  [file](https://github.com/openenclave/openenclave/blob/98b71a493cbbadda84c72187e9b5a59a3a1d7dec/samples/attested_tls/non_enc_client/client.cpp#L139 ). Is it IP-version agnostic? @dthaler